File: project/stack.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type sNode struct {
	person string
	next   *sNode
}

var gSTop *sNode

func sClear() {
	for gSTop != nil {
		gSTop = gSTop.next
	}
}

func sPush(v string) { gSTop = &sNode{person: v, next: gSTop} }

func sPop(out *string) bool {
	if gSTop == nil {
		return false
	}
	n := gSTop
	if out != nil {
		*out = n.person
	}
	gSTop = n.next
	return true
}

func sLoad(file string) bool {
	sClear()
	f, err := os.Open(file)
	if err != nil {
		return false
	}
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		sPush(s.Text())
	} // низ->верх, как в исходнике
	return true
}

func sSave(file string) bool {
	// сохранить низ->верх: соберём список и выведем в обратном порядке
	var arr []string
	for p := gSTop; p != nil; p = p.next {
		arr = append(arr, p.person)
	}
	f, err := os.Create(file)
	if err != nil {
		return false
	}
	defer f.Close()
	w := bufio.NewWriter(f)
	for i := len(arr) - 1; i >= 0; i-- {
		fmt.Fprintln(w, arr[i])
	}
	_ = w.Flush()
	return true
}

func sPrint() {
	first := true
	for p := gSTop; p != nil; p = p.next {
		if !first {
			fmt.Print(" ")
		}
		fmt.Print(p.person)
		first = false
	}
	fmt.Println()
}

File: stack.cpp
#include <iostream>
#include <string>
#include <fstream> 
#include <vector>
using namespace std;

struct SNode{
    string person;
    SNode* next;
};

SNode* newNode(const string& v, SNode* next = nullptr){
    return new SNode{v, next};
}

void print(SNode* top){ // печать стека сверху вниз
    cout << "------Стек------" << endl;
    int i = 0;
    while(top){
        cout << i++ << ") " << top->person << endl;
        top = top->next;
    }
}

size_t size(SNode* top){
    size_t n = 0;
    for(; top; top = top->next) ++n;
    return n;
}

bool empty(SNode* top){ return top == nullptr; }

SNode* peek(SNode* top){ // вершина (nullptr если пуст)
    return top;
}

void push(SNode** top, const string& v){
    *top = newNode(v, *top);
}

bool pop(SNode** top, string* out = nullptr){
    if(*top == nullptr) return false;
    SNode* t = *top;
    if(out) *out = t->person;
    *top = (*top)->next;
    delete t;
    return true;
}

void clear(SNode** top){
    string tmp;
    while(pop(top, &tmp)){}
}

static SNode* g_S_top = nullptr;

bool S_load(const string& file){
    clear(&g_S_top);
    ifstream f(file);
    if(!f) return false;
    // читаем низ->верх: просто пушим по порядку
    string s;
    while(getline(f, s)) push(&g_S_top, s);
    return true;
}
bool S_save(const string& file){
    ofstream f(file);
    if(!f) return false;
    // сохраняем низ->верх: соберём и выведем в обратном порядке
    vector<string> v;
    for(SNode* p=g_S_top; p; p=p->next) v.push_back(p->person);
    for(int i=(int)v.size()-1; i>=0; --i) f << v[i] << "\n";
    return true;
}
void S_push(const string& v){ push(&g_S_top, v); }
bool S_pop (string* out){    return pop(&g_S_top, out); }
void S_print(){
    bool first = true;
    for(SNode* p=g_S_top; p; p=p->next){
        if(!first) cout << ' ';
        cout << p->person;
        first = false;
    }
    cout << "\n";
}

#ifndef DS_SKIP_ONE_MAIN
int main(){
    SNode* top = nullptr;
    push(&top, "Первый");
    push(&top, "Второй");
    push(&top, "Третий");
    push(&top, "Четвертый");

    print(top);
    cout << "Размер: " << size(top) << endl;
    cout << "Пустой: " << (empty(top) ? "да" : "нет") << endl;

    cout << "Верх: " << (peek(top) ? peek(top)->person : string("null")) << endl;

    string val;
    if(pop(&top, &val)) cout << "Снято: " << val << endl;
    if(pop(&top, &val)) cout << "Снято: " << val << endl;

    print(top);
    cout << "Размер: " << size(top) << endl;

    clear(&top);
    cout << "Пустой после очистки: " << (empty(top) ? "да" : "нет") << endl;
    return 0;
}
#endif